### ISSUE TITLE
Dart2JS crashes when using pub from git

### ISSUE THREAD

**Issue by lexaknyazev** [dart-lang/sdk#30767]
Dart SDK 1.25.0-dev.16.4
1. Create bare-bones web app.
2. Run `pub build`. Everything builds normally:
```
Loading source assets...
Loading dart_to_js_script_rewriter transformers...
Building pub_bug...
[Info from Dart2JS]:
Compiling pub_bug|web/main.dart...
[Info from Dart2JS]:
Took 0:00:03.432658 to compile pub_bug|web/main.dart.
Built 240 files to "build".
```
3. Add this to `dev_dependencies`:
```
pub:
  git: https://github.com/dart-lang/pub.git
```
4. Run `pub get`.
5. Run `pub run pub build`. Dart2JS crashes:
```
Loading source assets...
Loading dart_to_js_script_rewriter transformers...
Building pub_bug...
[Info from Dart2JS]:
Compiling pub_bug|web/main.dart...
[Error from Dart2JS]:
web\main.dart:
The compiler crashed when compiling this element.
The compiler is broken.

When compiling the above element, the compiler crashed. It is not
possible to tell if this is caused by a problem in your program or
not. Regardless, the compiler should not crash.

The Dart team would greatly appreciate if you would take a moment to
report this problem at http://dartbug.com/new.

Please include the following information:

* the name and version of your operating system,

* the Dart SDK build number (build number could not be determined), and

* the entire message you see here (including the full stack trace
  below as well as the source location above).

Build error:
Transform Dart2JS on pub_bug|web/main.dart threw error: The getter 'isStatic' was called on null.
Receiver: null
Tried calling: isStatic
dart:core                                                         Object.noSuchMethod
package:compiler_unsupported/src/universe/world_builder.dart 744  ResolutionWorldBuilderImpl.registerStaticUse.<fn>
dart:collection                                                   _HashVMBase&MapMixin&&_LinkedHashMapMixin.putIfAbsent
package:compiler_unsupported/src/universe/world_builder.dart 743  ResolutionWorldBuilderImpl.registerStaticUse
package:compiler_unsupported/src/enqueue.dart 254                 ResolutionEnqueuer.processStaticUse
package:compiler_unsupported/src/enqueue.dart 467                 TreeShakingEnqueuerStrategy.processStaticUse
package:compiler_unsupported/src/enqueue.dart 509                 EnqueuerImplImpactVisitor.visitStaticUse
dart:collection                                                   _HashFieldBase&_HashBase&_OperatorEqualsAndHashCode&SetMixin.forEach
package:compiler_unsupported/src/util/setlet.dart 224             Setlet.forEach
package:compiler_unsupported/src/universe/world_impact.dart 39    WorldImpact.apply
package:compiler_unsupported/src/universe/world_impact.dart 268   ImpactStrategy.visitImpact
package:compiler_unsupported/src/enqueue.dart 180                 ResolutionEnqueuer.applyImpact
package:compiler_unsupported/src/compiler.dart 649                Compiler.compileLoadedLibraries.<fn>
package:compiler_unsupported/src/common/tasks.dart 176            CompilerTask.measureSubtask
package:compiler_unsupported/src/compiler.dart 603                Compiler.compileLoadedLibraries
package:compiler_unsupported/src/compiler.dart 511                Compiler.runInternal.<fn>
Build failed.
```

**Comment by sigmundch** [comment]
It is possible that this could be simply a bug that has already been fixed. When you use pub from a package, you end up using and old copy of dart2js that was published under compiler_unsupported (I believe it is currently 7 months old), but when you pub from the sdk, you also get the latest version of dart2js as well.

I don't expect failures like these on simple programs, though. If you attach the example you used to cause this crash, we might be able to tell if this was indeed something we fixed in the past few months.

**Comment by lexaknyazev** [comment]
By "bare-bones web app", I mean this one (from templates):
```dart
import 'dart:html';

void main() {
  querySelector('#output').text = 'Your Dart app is running.';
}
```

The same crash occurs on real application, too.

**Comment by sigmundch** [comment]
Ah - I found what the issue is: there is a mismatch between the sdk-libraries and the version of dart2js (dart2js has internal dependencies on the structure of the dart sdk libraries, and if a library changes internally it can break the compiler).  We should detect this and provide a nice error message in the future, but today this manifests as a crash in the compiler. I just validated that if I use the old compiler with today's libraries I get the behavior you are seeing.

When you use the `pub` from git, you are changing the version of dart2js, but not the version of the sdk libraries to match it. To change that together, you'll need to download an SDK that matches the same version that pub is using, and provide a flag to dart2js in the pubspec to use that version of the sdk libraries (the `--library-root` flag).
